### PR TITLE
fix(searchBox): handling pasting event with contextual menu.

### LINKED ISSUE #467: Handle paste events in searchBox
As detected by @dustincoates in instant search demo, we should also handle paste events in instantsearch.js in our searchBox.

Right now if you copy paste with the mouse in the searchBox then it will not trigger a search.

@dustincoates If you want to take this one, feel free!

### PR DESCRIPTION
**Pull request by maxiloc**
fixes #467

### CODE REVIEW

**Comment by redox** [comment]
@maxiloc I think we figured a more advanced way to handle that in https://github.com/algolia/instant-search-demo/pull/20 mentioned by @vvo in the issue.

**Comment by maxiloc** [comment]
Ok cool, I am modifying that

**Comment by redox** [comment]
> Ok cool, I am modifying that

:+1: awesome

**Comment by vvo** [comment]
I am not sure there's an "input" magical event in pure DOM

**Comment by pixelastic** [comment]
Should be pretty-well supported:
http://caniuse.com/#feat=input-event

But be careful, the order in which `input`, `change`, `keypress`, `keyup` and `keydown` are fired is tricky.

**Comment by dustincoates** [comment]
@maxiloc If you want to see how we did it in the demo: https://github.com/algolia/instant-search-demo/pull/20/files

@vvo's suggestion works pretty well because it also supports cutting and deleting with the mouse, plus the case where someone might paste text on to the end of an existing query.

**Comment by vvo** [comment]
So I wonder, @pixelastic can't we just use 'input' event for everything (it would handle keyup, change, paste..).

Then the only missing part would be that on IE9, deleting would not produce a new search, we can fix that separately.

**Comment by maxiloc** [comment]
I rebased, the handling of event is a bit different than jquery but it works

**Comment by vvo** [comment]
> I rebased, the handling of event is a bit different than jquery but it works

Did you pushed? :D

**Comment by pixelastic** [comment]
> So I wonder, @pixelastic can't we just use 'input' event for everything (it would handle keyup, change, paste..). Then the only missing part would be that on IE9, deleting would not produce a new search, we can fix that separately.

Yep, that could be a good idea (separate issue). We should fallback to classical keypress for IE9

**Comment by maxiloc** [comment]
Fixed test

File: widgets/search-box/__tests__/search-box-test.js [hunk 1]
```diff
@@ -182,7 +182,7 @@ describe('search-box()', () => {
         widget.getInput = sinon.stub().returns(input);
       });
 
-      it('performs the search on keyup', () => {
+      it('do not perform the search on keyup', () => {
```
Comment (redox) [inline widgets/search-box/__tests__/search-box-test.js]: I think the name of the test is not reflecting what it does. It definitely must "perform the search on keyup", right?

Comment (maxiloc) [inline widgets/search-box/__tests__/search-box-test.js]: yes we should not perform the search on key up. I will add a test for input event

**Comment by maxiloc** [comment]
added test for input event

**Comment by maxiloc** [comment]
This one is ready

File: widgets/search-box/__tests__/search-box-test.js (hunk 1 above)
Comment (vvo) [inline widgets/search-box/__tests__/search-box-test.js]: Are we good here? Not sure what @redox said was solved?

File: widgets/search-box/__tests__/search-box-test.js [hunk 2]
```diff
@@ -168,6 +168,59 @@ describe('search-box()', () => {
...
+      input.addEventListener = sinon.spy();
+    });
+
+    context('instant search', () => {
+      beforeEach(() => {
+        widget = searchBox({container, autofocus: 'auto'});
+        widget.getInput = sinon.stub().returns(input);
+      });
+
+      it('do not perform the search on keyup', () => {
```
Comment (pixelastic) [inline widgets/search-box/__tests__/search-box-test.js]: I feel like this test is also not correctly named. It simply tests that we bind an event listener to `input` and tests that event listener.

File: widgets/search-box/__tests__/search-box-test.js [hunk 3]
```diff
@@ -168,6 +168,59 @@ describe('search-box()', () => {
...
+        // Given
+        helper.state.query = 'foo';
+        // When
+        widget.init(initialState, helper);
+        // Then
+        expect(input.addEventListener.called).toEqual(true);
+        expect(input.addEventListener.args[1].length).toEqual(3);
+        expect(input.addEventListener.args[1][0]).toEqual('input');
```
Comment (pixelastic) [inline widgets/search-box/__tests__/search-box-test.js]: Shouldn't we do two tests, one for `addEventListener` and one for `attachEvent`, based on what the browser is able to do?

Those tests should only test that when triggering an `input` event on the search input, the correct method is called.

File: widgets/search-box/__tests__/search-box-test.js [hunk 4]
```diff
@@ -168,6 +168,59 @@ describe('search-box()', () => {
...
+        let fn = input.addEventListener.args[1][1];
+        fn({currentTarget: {value: 'test'}});
+        expect(helper.setQuery.calledOnce).toBe(true);
+        expect(helper.search.called).toBe(false);
```
Comment (pixelastic) [inline widgets/search-box/__tests__/search-box-test.js]: The only difference between the two tests is that `helper.search` is not called when `searchOnEnterKeyPressOnly: true`, still the test is very long. Shouldn't we just do a smaller test for that?

I find this two tests hard to follow

**Comment by redox** [comment]
I'll do a additional commit and merge, nice work @maxiloc :+1:

**Comment by vvo** [comment]
@maxiloc some comments from @pixelastic

**Comment by maxiloc** [comment]
@vvo I misunderstood the comment of @redox. I am gonna make the change

**Comment by maxiloc** [comment]
@pixelastic Rebased. Tests should be better

File: widgets/search-box/search-box.js [hunk 5]
```diff
@@ -100,13 +100,28 @@ function searchBox({
...
       });
 
+      function inputCallback(e) {
+        helper.setQuery(e.currentTarget.value);
+        if (!searchOnEnterKeyPressOnly) {
+          helper.search();
+        }
+      }
+
+      if (window.addEventListener) {
```
Comment (vvo) [inline widgets/search-box/search-box.js]: Can you reverse the test to (if attachEvent) else addeventlistener?

I double checked and it will support 9. IE9 has both addEventListener and attachEvent but the "input" event in IE9 will not be triggered when deleting text. While onpropertychange will. And propertychange must be done on attachEvent.

Double check that it works with a real IE9/10 (which have both attachEvent and addEventListener) because this is a tricky change.

Also check that it does not trigger too much the event, it seems that "propertychange" is triggered for every property change on the input. **tricky** :)

**Comment by maxiloc** [comment]
handled ie compatibility, we just won't have copy paste for ie9 for now

cc @vvo

**Comment by pixelastic** [comment]
I think you messed up something with the rebasing :)

image

**Comment by maxiloc** [comment]
weird :smile:

**Comment by maxiloc** [comment]
fixed

File: widgets/search-box/search-box.js [hunk 6]
```diff
@@ -100,13 +101,33 @@ function searchBox({
 
       // Add all the needed attributes and listeners to the input
       this.addDefaultAttributesToInput(input, initialState.query);
+
+      // Keep keyup to handle searchOnEnterKeyPressOnly
       input.addEventListener('keyup', (e) => {
         helper.setQuery(input.value);
-        if (!searchOnEnterKeyPressOnly || e.keyCode === KEY_ENTER) {
+        if (searchOnEnterKeyPressOnly && e.keyCode === KEY_ENTER) {
```
Comment (vvo) [inline widgets/search-box/search-box.js]: Can you add a comment stating that this is only needed in IE8/9? Thanks.

File: widgets/search-box/search-box.js [hunk 7]
```diff
@@ -100,13 +101,33 @@ function searchBox({
...
+      function inputCallback(e) {
+        let target = (e.currentTarget) ? e.currentTarget : e.srcElement;
+        helper.setQuery(target.value);
+        if (!searchOnEnterKeyPressOnly) {
+          helper.search();
+        }
+      }
+
+      if (window.attachEvent) {
+        input.attachEvent('onpropertychange', inputCallback);
```
Comment (vvo) [inline widgets/search-box/search-box.js]: Same here, can you add a comment stating this is only for IE8/9, they only support onpropertychange not the input event.

**Comment by maxiloc** [comment]
done

File: widgets/search-box/search-box.js (hunk 6 above)
Comment (vvo) [inline widgets/search-box/search-box.js]: cc @maxiloc

**Comment by vvo** [comment]
We are good, awesome work because it was a tricky one. If we have 0 bugs following this one then congrats :D
